### PR TITLE
feat(portex): support array unpack grammar of Portex

### PR DESCRIPTION
**Pull request by Lee-000**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2h0zen3 Portex schema 支持列表解包语法](https://app.clickup.com/t/2h0zen3)

File: graviti/portex/factory.py [hunk 1]
```diff
@@ -392,7 +400,10 @@ def __call__(self, **kwargs: Any) -> List[Any]:
             The applied list.
 
         """
-        return list(factory(**kwargs) for factory in self._factories)
+        array = list(factory(**kwargs) for factory in self._factories)
+        for factory in self._unpack_factories:
+            array.extend(factory(**kwargs))
```
Comment (linjiX) [inline graviti/portex/factory.py:405]: which means the unpacked object will be put at the end of the array?

File: graviti/portex/factory.py [hunk 2]
```diff
@@ -372,12 +372,20 @@ def __init__(self, decl: List[Any], ptype: PTYPE.PType = PTYPE.Any) -> None:
         dependences = set()
         keys = {}
         ptype = DynamicListParameter(ptype) if isinstance(ptype, Dynamic) else ptype
+        unpack_factories = []
         for value in decl:
+            if isinstance(value, str) and value.startswith("+$params."):
+                factory: Factory = array_unpack_factory_creator(value, PTYPE.Array)
```
Comment (linjiX) [inline graviti/portex/factory.py:378]: check "+$params." twice?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py (hunk 1 above)
Comment (Lee-000) [reply graviti/portex/factory.py:405]: Is it necessary to follow the order here?

Comment (linjiX) [reply graviti/portex/factory.py:405]: Of course the order matters

File: graviti/portex/factory.py [hunk 3]
```diff
@@ -545,6 +577,26 @@ def mapping_unpack_factory_creator(decl: str, ptype: PTYPE.PType) -> VariableFac
     return VariableFactory(decl[8:], ptype)
 
 
+def array_unpack_factory_creator(decl: str, ptype: PTYPE.PType) -> VariableFactory:
```
Comment (linjiX) [inline graviti/portex/factory.py:580]: do we still need these method?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 4]
```diff
@@ -616,8 +640,12 @@ def factory_creator(
         A ``Factory`` instance according the input.
 
     """
-    if isinstance(decl, str) and decl.startswith("$params."):
-        return VariableFactory(decl[8:], ptype)
+    if isinstance(decl, str):
+        if decl.startswith("$params."):
+            return VariableFactory(decl[8:], ptype)
+
+        if decl.startswith("+$params."):
```
Comment (linjiX) [inline graviti/portex/factory.py:647]: `TypeFactory` and `DictFactory` also use `factory_creator`, what if someone use `"+$params."` in `TypeFactory` and `DictFactory`?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/factory.py [hunk 5]
```diff
@@ -521,15 +539,28 @@ def __call__(self, **kwargs: Any) -> Fields:
...
         return Fields(
-            filter(bool, (factory(**kwargs) for factory in self._factories)),  # type: ignore[misc]
+            filter(bool, _generate_factory_items(self._factories, kwargs, lambda x: x.items()))
         )
 
 
+def _generate_factory_items(
+    factories: List[Factory],
+    kwargs: Dict[str, Any],
+    unpack_item_processer: Callable[[Any], Any] = lambda x: x,
```
Comment (linjiX) [inline graviti/portex/factory.py:549]: should it be `Callable[[Any], Iterable[Any]]`?

File: graviti/portex/factory.py [hunk 6]
```diff
@@ -521,15 +539,28 @@ def __call__(self, **kwargs: Any) -> Fields:
 
         """
         return Fields(
-            filter(bool, (factory(**kwargs) for factory in self._factories)),  # type: ignore[misc]
+            filter(bool, _generate_factory_items(self._factories, kwargs, lambda x: x.items()))
         )
 
 
+def _generate_factory_items(
+    factories: List[Factory],
```
Comment (linjiX) [inline graviti/portex/factory.py:547]: `Iterable[Factory]`?

**Review by linjiX — APPROVED** [review]

**Pull request merged**